TypeCobol/Compiler/Preprocessor/CopyWrappedAsProgramTokensDocument.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using TypeCobol.Compiler.Concurrency;
+using TypeCobol.Compiler.Directives;
+using TypeCobol.Compiler.Parser;
+using TypeCobol.Compiler.Scanner;
+using TypeCobol.Compiler.Text;
+
+namespace TypeCobol.Compiler.Preprocessor
+{
+    /// <summary>
+    /// Special ProcessedTokensDocument that wraps original lines of a copy into a fake program
+    /// to allow full parsing and analysis on copies.
+    /// </summary>
+    public class CopyWrappedAsProgramTokensDocument : ProcessedTokensDocument
+    {
+        private static ISearchableReadOnlyList<CodeElementsLine> Wrap(ISearchableReadOnlyList<CodeElementsLine> originalLines, TextSourceInfo textSourceInfo)
```

Comment:
I'm not conviced by this wrapping of CodeElementsLine.
It breaks the logic of the parser.
Lines defined in CompilationDocument are not supposed to be altered, hence the use of an ISearchableReadOnlyList.


Is it not possible to use the mechanism of Iterator (like AutoReplacePartialWordsTokensLinesIterator) to return all tokens generated in this class ?
I mean a new iterator on top of AutoReplacePartialWordsTokensLinesIterator. This new iterator will :
- first returns token before the copy content (id division, program Id, data division, working-storage, root 01).
- returns the content of the copy
- returns END PROGRAM tokens.
I think we've already talked about this solution.


This mimic the behavior of copy inclusion without altering Lines of CompilationDocument.

Reply:
Another problem if lines are modified: On LSP, we use differents kinds of object to retrieve Lines:
- CompilationDocument.CobolTextLines
- ProcessedTokensDocument.Lines
- CodeElementsDocument.Lines

Depending on the case, we will have generated CodeElement or not.
For example `OnHover` will detect `IDENTIFICATION DIVISION` CodeElement if you hover on the first line of the COPY.


After secod thought, you could use `CodeElementTokenizer `to return generated CodeElement instead of generating Token.
This way you don't have to worry about Incremental mode and CodeElement will not be attached to Lines.

Reply:
Done in 967cef7e9ab7ebe1b99961ffab15748017bb9b5a.

Reply:
After a quick test, it works fiine now with LSP.
OnHover now detect the correct CodeElement.